### ISSUE TITLE
when LinearReplacer is used?

### ISSUE THREAD

**Issue by 191220042** [Azure/MS-AMP#197]
Hello,when I read the code, I found that no matter FP_TYPE is te or msamp, the parameter transformer_engine will be used. So when will TeReplacer.replace be used?
![Image](https://github.com/user-attachments/assets/f7127771-d614-4fd1-964b-ad3ac0beb8f7)

**Comment by Mr-Philo** [comment]
Hi @191220042 ,
When `args.transformer_impl == 'transformer_engine'`, the model constructer will use fused liner layer (written by TE) instead of torch.nn.linear layer, thus LinearReplacer won't be effective.
To inspect the effect of LinearReplacer, please modify this argparse in your python command:
```sh
--transformer-impl "local"
```
Then you should see the effect of FP8Linear! For double check, you can also consider printing the model structure after LinearReplacer is called.

**Comment by 191220042** [comment]
> Hi [@191220042](https://github.com/191220042) , When `args.transformer_impl == 'transformer_engine'`, the model constructer will use fused liner layer (written by TE) instead of torch.nn.linear layer, thus LinearReplacer won't be effective. To inspect the effect of LinearReplacer, please modify this argparse in your python command:
> 
> --transformer-impl "local"
> Then you should see the effect of FP8Linear! For double check, you can also consider printing the model structure after LinearReplacer is called.

Thank you for your reply, but I have tried such changes before, and local conflicts with use_fp8.![Image](https://github.com/user-attachments/assets/074eb441-b7c0-479a-9a9e-52edc69f8441)

**Comment by Mr-Philo** [comment]
Oh I see that you're using the script from [MS-AMP-Example](https://github.com/Azure/MS-AMP-Examples/tree/main/gpt3), which provided a training shell script like this:

```sh
if [ "$FP_TYPE" = "bf16" ]; then
    CHECKPOINT_PATH=$PWD/checkpoints/gpt_345m_bf16
    torchrun $DISTRIBUTED_ARGS ../third_party/Megatron-LM/pretrain_gpt.py \
        $GPT_ARGS \
        $DATA_ARGS \
        $OUTPUT_ARGS \
        --save $CHECKPOINT_PATH \
        --load $CHECKPOINT_PATH

elif [ "$FP_TYPE" = "te" ]; then
    CHECKPOINT_PATH=$PWD/checkpoints/gpt_345m_te
    torchrun $DISTRIBUTED_ARGS ../third_party/Megatron-LM/pretrain_gpt.py \
        $GPT_ARGS \
        $DATA_ARGS \
        $OUTPUT_ARGS \
        --fp8-hybrid \
        --transformer-impl transformer_engine \
        --save $CHECKPOINT_PATH \
        --load $CHECKPOINT_PATH

elif [ "$FP_TYPE" = "msamp" ]; then
    CHECKPOINT_PATH=$PWD/checkpoints/gpt_345m_msamp
    torchrun $DISTRIBUTED_ARGS ../third_party/Megatron-LM/pretrain_gpt.py \
        $GPT_ARGS \
        $DATA_ARGS \
        $OUTPUT_ARGS \
        --fp8-hybrid \
        --transformer-impl transformer_engine \
        --msamp \
        --save $CHECKPOINT_PATH \
        --load $CHECKPOINT_PATH
```

This script provided three levels of training: BF16, TE, TE+**fused msamp**. In the third level, MS-AMP was fused into the TE traininig scheme. Under this setting, `TeReplacer` instead of `LinearReplacer` will be used.
If you want to inspect the effect of `LinearReplacer` alone, you could delete the `--fp8-hybrid` option since this option is controling whether to open FP8 TE support in Megatron structure, but has no relationship with MS-AMP itself. In other words, launching the script like this:

```sh
CHECKPOINT_PATH=$PWD/checkpoints/gpt_345m_msamp_alone
torchrun $DISTRIBUTED_ARGS ../third_party/Megatron-LM/pretrain_gpt.py \
        $GPT_ARGS \
        $DATA_ARGS \
        $OUTPUT_ARGS \
        --transformer-impl local \
        --msamp \
        --save $CHECKPOINT_PATH \
        --load $CHECKPOINT_PATH
```
will enable `LinearReplacer` and use MS-AMP FP8 training, no TE FP8 or TE-MS-AMP-fused FP8 training.

**Comment by 191220042** [comment]
hello,

**Comment by 191220042** [comment]
> Oh I see that you're using the script from [MS-AMP-Example](https://github.com/Azure/MS-AMP-Examples/tree/main/gpt3), which provided a training shell script like this:
> 
> if [ "$FP_TYPE" = "bf16" ]; then
>     CHECKPOINT_PATH=$PWD/checkpoints/gpt_345m_bf16
>     torchrun $DISTRIBUTED_ARGS ../third_party/Megatron-LM/pretrain_gpt.py \
>         $GPT_ARGS \
>         $DATA_ARGS \
>         $OUTPUT_ARGS \
>         --save $CHECKPOINT_PATH \
>         --load $CHECKPOINT_PATH
> 
> elif [ "$FP_TYPE" = "te" ]; then
>     CHECKPOINT_PATH=$PWD/checkpoints/gpt_345m_te
>     torchrun $DISTRIBUTED_ARGS ../third_party/Megatron-LM/pretrain_gpt.py \
>         $GPT_ARGS \
>         $DATA_ARGS \
>         $OUTPUT_ARGS \
>         --fp8-hybrid \
>         --transformer-impl transformer_engine \
>         --save $CHECKPOINT_PATH \
>         --load $CHECKPOINT_PATH
> 
> elif [ "$FP_TYPE" = "msamp" ]; then
>     CHECKPOINT_PATH=$PWD/checkpoints/gpt_345m_msamp
>     torchrun $DISTRIBUTED_ARGS ../third_party/Megatron-LM/pretrain_gpt.py \
>         $GPT_ARGS \
>         $DATA_ARGS \
>         $OUTPUT_ARGS \
>         --fp8-hybrid \
>         --transformer-impl transformer_engine \
>         --msamp \
>         --save $CHECKPOINT_PATH \
>         --load $CHECKPOINT_PATH
> This script provided three levels of training: BF16, TE, TE+**fused msamp**. In the third level, MS-AMP was fused into the TE traininig scheme. Under this setting, `TeReplacer` instead of `LinearReplacer` will be used. If you want to inspect the effect of `LinearReplacer` alone, you could delete the `--fp8-hybrid` option since this option is controling whether to open FP8 TE support in Megatron structure, but has no relationship with MS-AMP itself. In other words, launching the script like this:
> 
> CHECKPOINT_PATH=$PWD/checkpoints/gpt_345m_msamp_alone
> torchrun $DISTRIBUTED_ARGS ../third_party/Megatron-LM/pretrain_gpt.py \
>         $GPT_ARGS \
>         $DATA_ARGS \
>         $OUTPUT_ARGS \
>         --transformer-impl local \
>         --msamp \
>         --save $CHECKPOINT_PATH \
>         --load $CHECKPOINT_PATH
> will enable `LinearReplacer` and use MS-AMP FP8 training, no TE FP8 or TE-MS-AMP-fused FP8 training.

hello,When I follow the local configuration, I notice that the ddp takes the <class 'msamp.megatron.distributed.FP8DistributedDataParallel' >And this class directly inherits the DistributedDataParallelBase of megatron.model.distributed, and I'm curious about where the logic of each rank gradient allreduce and communication scale is different from the FP8DistributedDataParallel custom in msamp's nn你好，当我按照local的配置时，我注意到ddp会采用<class 'msamp.megatron.distributed.FP8DistributedDataParallel'>而这个类直接继承megatron.model.distributed的DistributedDataParallelBase，我很好奇与msamp的nn中自定义的FP8DistributedDataParallel不同，在local情况下各个rank梯度allreduce以及通信scale的逻辑在什么地方![Image](https://github.com/user-attachments/assets/b659a51c-2cf3-4ab6-acd7-f43d0cd565de)

**Comment by wkcn** [comment]
Hi @191220042, thanks for your attention to our work!

All reduce of FP8 weight gradient: https://github.com/Azure/MS-AMP/blob/main/msamp/megatron/optimizer/distrib_optimizer.py#L796-L804
Communication of scaling factor：https://github.com/Azure/MS-AMP/blob/main/msamp/megatron/optimizer/distrib_optimizer.py#L777-L783
